CRM/Utils/System/Standalone.php

```diff
@@ -245,6 +245,20 @@ public function logout() {
     return Security::singleton()->logoutUser();
   }
 
+  /**
+   * @inheritDoc
+   *
+   */
+  public static function getContentTemplate(bool $print = FALSE, bool $frontEnd = FALSE): string {
+    if ($print) {
+      return 'CRM/common/print.tpl';
+    }
+    $isPublic = \CRM_Core_Smarty::singleton()->getTemplateVars('urlIsPublic');
+    // Alternative:
+    // $isPublic = ($_GET['q'] ?? '') ? CRM_Core_Menu::isPublicRoute($_GET['q']) : FALSE;
```

Comment:
I think my pref would be for the CRM_Core_Menu::isPublicRoute option... feels like controller logic that belongs above Smarty maybe? and more hookable?

Reply:
I worry a bit about someone coming along and calling this function before the smarty vars have been assigned though?

Could we pass the value of the smarty var in using the frontend param, and then just use the explicitly passed in value?

Reply:
was trying to understand how this works generally and found third option `CRM_Utils_System::isFrontEndPage` 

I don't really understand how the implementation in the Base system class works though. maybe we should be taking the DrupalBase implementation into the Standalone system?

Reply:
looking again, I think the base implementation is just always return FALSE?

Reply:
Yeah, there are multiple ways to discern frontend from backend. That's fun.

* `CRM_Utils_System::isFrontEndPage()`  (aka `$config->userFrameworkFrontend`)
    * It's also used for CiviCRM's theming. Remember there are two settings: `theme_backend` and `theme_frontend`. This method decides which theme to use.
    * It's also used if you construct URL's like `Civi::url('current://civicrm/profile/view?id=123')`. (The `current://` is interpreted as `frontend://` or `backend://` depending on the value of `isFrontEndPage()`).
    * This is set on Joomla and WordPress. It it is unset on Drupal/Backdrop.
* `CRM_Core_Menu::isPublicRoute()` (aka Smarty `$urlIsPublic`)
    * Many/most routes have a singular presentation (frontend XOR backend). So you can *usually* use the routing table.
    * There are some routes like `civicrm/profile/*` which can reasonably be used in fronted AND/OR backend. (__Frontend__: "New subscriber signup." "Public partner directory." __Backend__: "Update your account." "Staff directory.")

The thing that sticks out to me - if there are two HTML templates ("frontend/backend") and two CSS themes ("frontend/backend"), then they should match-up ("frontend HTML <===> frontend CSS"). So using `isFrontEndPage()` seems more sensible.

But AFAICS, Standalone doesn't populate `$config->userFrameworkFrontend` - so actually,  right now, it might not give any useful signal. I think that:

1. If standalone is to have a frontend/backend split, then it needs to set `userFrameworkFrontend` to ensure that CSS theming works and matches.
2. The question will be standalone's source-of-truth, of which I know two options:
     * __Static/Routing Table/Single Truth__: You read from the routing table and use that to decide if the route is frontend/backend.
         * This means that, eg, `civicrm/profile/*` cannot be multi-homed. It has exactly one home.
         * In this case, Standalone should read `CRM_Core_Menu::isPublicRoute()` and set`userFrameworkFrontend`.
     * __Dynamic/Request Properties/Multiple Truths__: You look at the URL of the request, like J/WP (e.g. `/index.php` vs `/admin/index.php`). 
         * This means that `civicrm/profile/*` has both frontend+backend flavors.
         * In this case, Standalone needs another entry-point... (or some other URL-based signal...)

Reply:
(1) Aside: Ooh, missed something. Drupal already implements [isFrontEndPage()](https://github.com/civicrm/civicrm-core/blob/f5119f130ff6d51268e3a82e6eea5d16c49c83e3/CRM/Utils/System/DrupalBase.php#L716-L723). And it is (*substantively*) doing the static route-lookup.(*The code predates `isPublicRoute()` so it doesn't actually call that.*)

(2) Suggested patch:

```diff
diff --git a/CRM/Utils/System/Standalone.php b/CRM/Utils/System/Standalone.php
index 220fe3471d..7efda80cbc 100644
--- a/CRM/Utils/System/Standalone.php
+++ b/CRM/Utils/System/Standalone.php
@@ -253,9 +253,7 @@ class CRM_Utils_System_Standalone extends CRM_Utils_System_Base {
     if ($print) {
       return 'CRM/common/print.tpl';
     }
-    $isPublic = \CRM_Core_Smarty::singleton()->getTemplateVars('urlIsPublic');
-    // Alternative:
-    // $isPublic = ($_GET['q'] ?? '') ? CRM_Core_Menu::isPublicRoute($_GET['q']) : FALSE;
+    $isPublic = CRM_Utils_System::isFrontEndPage();
     return $isPublic ? 'CRM/common/standalone-frontend.tpl' : 'CRM/common/standalone.tpl';
   }
 
@@ -386,6 +384,10 @@ class CRM_Utils_System_Standalone extends CRM_Utils_System_Base {
     throw new \RuntimeException("Standalone requires the path is set for now. Set \$civicrm_paths['cms.root']['path'] in civicrm.settings.php to the webroot.");
   }
 
+  public function isFrontEndPage() {
+    return CRM_Core_Menu::isPublicRoute(CRM_Utils_System::currentPath() ?? '');
+  }
+
   /**
    * @inheritDoc
    */
```

That means HTML template and CSS theme and URL generator will use the same signal about whether the current page-request is frontend. And that signal will be based on the static routing table.

(3) Re: "profile front end pages should be given a 2nd route"... I think we can push that question to some other issue/time. Figuring out exactly what to call those routes would be... fiddly. And I don't think a change on that is pressing.